src/rust/generator-dispatcher/src/lib.rs

```diff
@@ -82,13 +100,81 @@ impl GeneratorDispatcher {
+        let client_config = PluginRegistryClientConfig::parse();
+        let plugin_registry_client = build_grpc_client(client_config).await?;
+        let generator_ids_cache = AsyncCache::new(
             config.params.generator_ids_cache_capacity,
             Duration::from_millis(config.params.generator_ids_cache_ttl_ms),
             config.params.generator_ids_cache_updater_pool_size,
             config.params.generator_ids_cache_updater_queue_depth,
+            move |event_source_id| {
+                let mut plugin_registry_client = plugin_registry_client.clone();
+
+                async move {
+                    match plugin_registry_client
+                        .get_generators_for_event_source(GetGeneratorsForEventSourceRequest::new(
+                            event_source_id,
+                        ))
+                        .await
+                    {
+                        Ok(response) => response.plugin_ids().to_vec(),
+                        Err(GrpcClientError::ErrorStatus(Status {
+                            code: Code::NotFound,
+                            ..
+                        })) => {
+                            tracing::warn!(
+                                message = "found no generators for event source",
+                                event_source_id =% event_source_id,
+                            );
+                            vec![]
+                        }
+                        Err(e) => {
+                            // received an error response from the
+                            // plugin-registry service, so we'll retry
+                            // indefinitely using a truncated binary
+                            // exponential backoff with jitter, capped
+                            // at 5s.
+
+                            let mut result: Result<
+                                GetGeneratorsForEventSourceResponse,
+                                GrpcClientError,
+                            > = Err(e);
+                            let mut n = 0;
+                            while let Err(ref e) = result {
+                                n += 1;
+                                let millis = 2_u64.pow(n)
+                                    + rand::thread_rng().gen_range(0..2_u64.pow(n - 1));
+                                let backoff = if millis < 5000 {
+                                    Duration::from_millis(millis)
+                                } else {
+                                    Duration::from_millis(10000)
+                                };
+
+                                tracing::error!(
+                                    message = "error retrieving generator IDs from plugin-registry",
+                                    error =% e,
+                                    retry_delay =? backoff,
+                                );
+
+                                tokio::time::sleep(backoff).await;
```

Comment:
A) `get_generators_for_event_source` already _does_ retries itself (See `macro_rules! execute_client_rpc`), so I'm a bit confused as to why we're adding another layer on top.
B) At the very least, I'd prefer to do this in a separate bare function than nested so deep here, if possible?

Reply:
I tried following that code through and couldn't figure it out. Where does the retry happen and how can I configure it?